exploration/0003-formatted-parts.md

```diff
@@ -0,0 +1,187 @@
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The parts' `dir` and `locale` may be defined if overridden by an expression attribute or set otherwise.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
+
+```ts
+interface MessageExpressionPart {
+  type: string;
+  source: string;
+  parts?: Iterable<{ type: string; value: unknown }>;
+  value?: unknown;
+  dir?: "ltr" | "rtl" | "auto";
+  locale?: string;
+}
+```
+
+The bidi isolation strategies included in the spec may require
+the insertion of MessageBiDiIsolationParts in the formatted-parts output.
+
+```ts
+interface MessageBiDiIsolationPart {
+  type: "bidiIsolation";
+  value: "\u2066" | "\u2067" | "\u2068" | "\u2069"; // LRI | RLI | FSI | PDI
+}
+```
+
+Some of the MessageExpressionPart instances may be further defined
+without reference to the function registry.
+
+Unannotated expressions with a _literal_ operand
+are represented by MessageStringPart.
+As with MessageTextPart,
+the `value` of MessageStringPart is always a string.
+
+```ts
+interface MessageStringPart {
+  type: "string";
+  source: string;
+  value: string;
+  dir?: "ltr" | "rtl" | "auto";
+  locale?: string;
+}
+```
+
+Unannotated expressions with a _variable_ operand
+that is not formatted according to its type
+are represented by MessageUnknownPart.
+
+```ts
+interface MessageUnknownPart {
+  type: "unknown";
+  source: string;
+  value: unknown;
+}
+```
+
+When the resolution or formatting of a placeholder fails,
+it is represented in the output by MessageFallbackPart.
+No `value` is provided; when formatting to a string,
+the part's representation would be `'{' + source + '}'`.
+
+```ts
+interface MessageFallbackPart {
+  type: "fallback";
+  source: string;
+}
+```
+
+Formatting functions defined in the registry
+will need accompanying formatted-parts representations.
+Where available, such a formatted value should itself be represented by `parts`
+rather than a unitary string `value`.
+These sub-parts should not need fields beyond their `type` and `value`,
+and in most cases it's presumed that the sub-part `value` would be a string.
+
+```ts
+interface MessageDateTimePart {
+  type: "datetime";
+  source: string;
+  parts: Iterable<{ type: string; value: unknown }>;
```

Comment:
Notice in my example earlier in this review that I exposed the field name as a field in the parts. This becomes important when trying to decorate an Iterable whose contents shift around due to the locale/localized formatting. Dates have this feature (YMD, DMY, MDY). So do currency values (which may or may not have a decimal part, may have the symbol first or last, and may or may not have a space around the symbol). That's how the screen shots of currency values (from amazon.com and amazon.fr) get decorated.

Reply:
Yes, that's the `"type"` field here. I picked that rather than `"name"` because it's used by the JS [Intl formatters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/formatToParts#description).

Reply:
That was super non-obvious to me (hence this conversation!), particularly since the `type` fields in the examples seemed to be focused on the "type" of formatter (datetime, number) rather than on the parts field within them. Admittedly, the MF2-level parts will be at the placeholder level. Interior parts are the problem of the formatter. But this was not at all clear and probably could use an example.

Reply:
Let's make sure that it's explained in more detail in the spec PR, should this design doc be accepted.